### PR TITLE
K3s Integration

### PR DESCRIPTION
**Pull request by daxmc99**
This adds the ability for Rancher to import k3s clusters and upgrade them

Addresses: https://github.com/rancher/rancher/issues/25371

### CODE REVIEW

File: pkg/controllers/management/k3supgrade/template.go [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+metadata:
+  name: k3s-master-plan 
+  namespace: system-upgrade
+spec:
+  concurrency: 1
+  version: v1.17.2-k3s1
+  nodeSelector:
+    matchExpressions:
+      - {key: k3s-upgrade, operator: Exists}
+  serviceAccountName: default
```
Comment (dweomer) [inline pkg/controllers/management/k3supgrade/template.go]:
```suggestion
  serviceAccountName: system-upgrade # or cattle-upgrade
```

Comment (daxmc99) [reply pkg/controllers/management/k3supgrade/template.go]: https://github.com/rancher/rancher/pull/25667/files#diff-f70d9c6e95a58b98b200416cd10ef05eR54 
These plans are actually unused in their current form

File: pkg/api/customization/kontainerdriver/actionhandler.go [hunk 2]
```diff
@@ -137,12 +137,12 @@ func (lh ListHandler) LinkHandler(apiContext *types.APIContext, next types.Reque
 	var targetImages []string
 	switch apiContext.ID {
 	case linuxImages:
-		targetImages, err = image.GetImages(systemCatalogChartPath, []string{}, rkeSysImages, image.Linux)
+		targetImages, err = image.GetImages(systemCatalogChartPath, []string{}, []string{}, rkeSysImages, image.Linux)
```
Comment (rmweir) [inline pkg/api/customization/kontainerdriver/actionhandler.go]: this may have been me, but I believe this was a mistake. Should be rkeSysimages still.

File: pkg/controllers/management/k3supgrade/register.go [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+		return err
+	}
+	planClient := planConfig.Plans(metav1.NamespaceAll)
+
+	planList, err := planClient.List(metav1.ListOptions{})
+	if err != nil {
+		// may need to handle this error, if there is no Plan CRD what should we do?
+		if errors.IsNotFound(err) {
+			// no plan CRD exists
+			logrus.Warnf("plan CRD does not exist: %s", err)
```
Comment (rmweir) [inline pkg/controllers/management/k3supgrade/register.go]: nit: maybe just make this the err that is returned in this case and say, the crd does not exist _yet_

File: pkg/controllers/management/k3supgrade/register.go [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+				return err
+			}
+			if !cmp(masterPlan, newMaster) {
+				planClient = planConfig.Plans(systemUpgradeNS)
+				_, err = planClient.Update(&newMaster)
+				if err != nil {
+					return err
+				}
+			} else {
+				fmt.Println("master plan is the same, not updating")
```
Comment (rmweir) [inline pkg/controllers/management/k3supgrade/register.go]: should this be an info log?

File: pkg/controllers/management/k3supgrade/register.go [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+				return err
+			}
+			if !cmp(workerPlan, newWorker) {
+				planClient = planConfig.Plans(systemUpgradeNS)
+				_, err = planClient.Update(&newWorker)
+				if err != nil {
+					return nil
+				}
+			} else {
+				fmt.Println("worker plan is the same, not updating")
```
Comment (rmweir) [inline pkg/controllers/management/k3supgrade/register.go]: same as L141

File: pkg/controllers/management/k3supgrade/register.go [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+			}
+		}
+
+	} else { // create the plans
+		planClient = planConfig.Plans(systemUpgradeNS)
+		masterPlan, err = generateMasterPlan(cluster.Spec.K3sConfig.Version,
```
Comment (rmweir) [inline pkg/controllers/management/k3supgrade/register.go]: do we not want to error check this?

File: pkg/controllers/management/k3supgrade/register.go [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+			cluster.Spec.K3sConfig.ServerConcurrency)
+		_, err = planClient.Create(&masterPlan)
+		if err != nil {
+			return err
+		}
+		workerPlan, err = generateWorkerPlan(cluster.Spec.K3sConfig.Version,
```
Comment (rmweir) [inline pkg/controllers/management/k3supgrade/register.go]: same as L163

File: pkg/controllers/management/k3supgrade/register.go [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+			cluster.Spec.K3sConfig.WorkerConcurrency)
+		_, err = planClient.Create(&workerPlan)
+		if err != nil {
+			return nil
+		}
+		fmt.Println("Deployed plans into cluster")
```
Comment (rmweir) [inline pkg/controllers/management/k3supgrade/register.go]: info log?

File: pkg/controllers/management/k3supgrade/template.go [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+     image: rancher/k3s-upgrade:latest
+     args: ["prepare","k3s-master-plan"]
+  nodeSelector:
+    matchExpressions:
+    - {key: k3s-worker-upgrade, operator: Exists}
+  serviceAccountName: system-upgrade
+  drain:
+    force: true
+  upgrade:
+    image: rancher/k3s-upgrade`
```
Comment (rmweir) [inline pkg/controllers/management/k3supgrade/template.go]: If we don't use these plan constants can we just delete them?

File: pkg/controllers/management/k3supgrade/template.go [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+		Upgrade: &planv1.ContainerSpec{
+			Image: upgradeImage,
+		},
+	},
+	Status: planv1.PlanStatus{},
+}
+
+func generateMasterPlan(version string, concurrency int) (planv1.Plan, error) {
+	masterPlan := genericPlan
+	masterPlan.Name = k3sMasterPlanName
```
Comment (rmweir) [inline pkg/controllers/management/k3supgrade/template.go]: do we need use selectors if we already know the name? and will there be an issue ever with us hardcoding the name? There should never be two?

File: pkg/controllers/management/k3supgrade/template.go [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+			Values:   []string{"true"},
+		}},
+	}
+
+	return masterPlan, nil
+}
+
+func generateWorkerPlan(version string, concurrency int) (planv1.Plan, error) {
+	workerPlan := genericPlan
+	workerPlan.Name = k3sWorkerPlanName
```
Comment (rmweir) [inline pkg/controllers/management/k3supgrade/template.go]: same as L86

File: pkg/image/export/main.go [hunk 12]
```diff
@@ -77,12 +77,14 @@ func run(systemChartPath string, imagesFromArgs []string) error {
...
-	targetImages, err := img.GetImages(systemChartPath, imagesFromArgs, linuxInfo.RKESystemImages, img.Linux)
+	k3sUpgradeImages := getK3sUpgradeImages(data.K3S)
+
+	targetImages, err := img.GetImages(systemChartPath, k3sUpgradeImages, imagesFromArgs, linuxInfo.RKESystemImages, img.Linux)
 	if err != nil {
 		return err
 	}
 
-	targetWindowsImages, err := img.GetImages(systemChartPath, []string{getWindowsAgentImage()}, windowsInfo.RKESystemImages, img.Windows)
+	targetWindowsImages, err := img.GetImages(systemChartPath, k3sUpgradeImages, []string{getWindowsAgentImage()}, windowsInfo.RKESystemImages, img.Windows)
```
Comment (rmweir) [inline pkg/image/export/main.go]: k3sUpgradeImages should probably be empty. Even though a windows cluster will be a mixture of windows and linux nodes, the windows-images textfile is just to explain the images needed for the windows nodes I believe, as the linux ones will be downloaded anyway. See: https://rancher.com/docs/rancher/v2.x/en/installation/other-installation-methods/air-gap/populate-private-registry/#linux-steps
Could have been me but I thought I changed this.

File: pkg/controllers/management/clusterprovisioner/provisioner.go [hunk 13]
```diff
@@ -966,10 +966,10 @@ func k3sClusterConfig(cluster *v3.Cluster) error {
 			Err:    fmt.Errorf("waiting for full cluster configuration"),
 			Reason: "Pending"}
 	}
-	// TODO Rancher may support upgrading the local cluster if its k3s in the future
-	if cluster.Name == v3.ClusterDriverLocal || cluster.Status.Driver == v3.ClusterDriverK3s {
-		return nil
+	if cluster.Status.Driver == v3.ClusterDriverK3s {
```
Comment (galal-hussein) [inline pkg/controllers/management/clusterprovisioner/provisioner.go]: I am not sure I understand why did you remove the exception for local clusters here

File: pkg/controllers/management/k3supgrade/template.go (hunk 10 above)
Comment (daxmc99) [reply pkg/controllers/management/k3supgrade/template.go]: Selectors are required to activate a plan (cc @dweomer) and we do not want to label every node to activate a plan. There should only ever be two _rancher-managed_ plans. One master and one worker

File: pkg/controllers/management/k3supgrade/register.go (hunk 3 above)
Comment (daxmc99) [reply pkg/controllers/management/k3supgrade/register.go]: I don't think it matters what the error here is actually. We should just requeue on any error.

Comment (daxmc99) [reply pkg/controllers/management/k3supgrade/register.go]: In other controllers we may not return the 'NotFound' error to the framework but here we know that the system-upgrade-controller should be installed

File: pkg/controllers/management/k3supgrade/register.go (hunk 4 above)
Comment (daxmc99) [reply pkg/controllers/management/k3supgrade/register.go]: Not sure if this is worth a log

File: pkg/controllers/management/k3supgrade/register.go (hunk 8 above)
Comment (daxmc99) [reply pkg/controllers/management/k3supgrade/register.go]: Same here, going to drop the log

File: pkg/api/customization/kontainerdriver/actionhandler.go (hunk 2 above)
Comment (daxmc99) [reply pkg/api/customization/kontainerdriver/actionhandler.go]: This is still `rkeSysImages`. Current func signature is `GetImages(systemChartPath string, k3sUpgradeImages, imagesFromArgs []string, rkeSystemImages map[string]v3.RKESystemImages, osType OSType) ([]string, error)` 
arg[3] is and should be rkeSysImage ✨

File: pkg/image/export/main.go [hunk 14]
```diff
@@ -218,6 +220,20 @@ func getWindowsAgentImage() string {
 	return fmt.Sprintf("%s/rancher-agent:%s", repo, tag)
 }
 
+func getK3sUpgradeImages(k3sData map[string]interface{}) []string {
```
Comment (cjellick) [inline pkg/image/export/main.go]: This function is fairly funky in what it does. I think it needs a comment

File: pkg/controllers/management/clusterprovisioner/provisioner.go (hunk 13 above)
Comment (daxmc99) [reply pkg/controllers/management/clusterprovisioner/provisioner.go]: @cjellick Asked that we be able to recognize local clusters if they are k3s. Do you see any issues here?

File: pkg/controllers/management/k3supgrade/template.go (hunk 9 above)
Comment (daxmc99) [reply pkg/controllers/management/k3supgrade/template.go]: done

File: pkg/controllers/management/k3supgrade/register.go (hunk 6 above)
Comment (daxmc99) [reply pkg/controllers/management/k3supgrade/register.go]: dropped error creation in `generateMasterPlan`

File: pkg/controllers/management/k3supgrade/k3supgrade_handler.go [hunk 15]
```diff
@@ -0,0 +1,174 @@
...
+
+// deployK3sUpgradeController creates a rancher k3s upgrader controller if one does not exist.
+// Updates k3s upgrader controller if one exists and is not the newest available version.
+func (h *handler) deployK3sUpgradeController(clusterName string) error {
+	userCtx, err := h.manager.UserContext(clusterName)
+	if err != nil {
+		return err
+	}
+
+	projectLister := userCtx.Management.Management.Projects("").Controller().Lister()
```
Comment (dramich) [inline pkg/controllers/management/k3supgrade/k3supgrade_handler.go]: This can just live on the handler so it doesn't need to be created here

File: pkg/controllers/management/k3supgrade/k3supgrade_handler.go [hunk 16]
```diff
@@ -0,0 +1,174 @@
...
+	systemProject, err := project.GetSystemProject(clusterName, projectLister)
+	if err != nil {
+		return err
+	}
+
+	templateID := k3sUpgraderCatalogName
```
Comment (dramich) [inline pkg/controllers/management/k3supgrade/k3supgrade_handler.go]: Why the new var?

File: pkg/controllers/management/k3supgrade/register.go [hunk 17]
```diff
@@ -0,0 +1,190 @@
...
+	templateLister         v3.CatalogTemplateLister
+	systemAccountManager   *systemaccount.Manager
+	manager                *clustermanager.Manager
+}
+
+const (
+	systemUpgradeNS        = "cattle-system"
+	rancherManagedPlan     = "rancher-managed"
+	upgradeDisableLabelKey = "plan.upgrade.cattle.io/disable"
+	k3sUpgraderCatalogName = "system-library-rancher-k3s-upgrader"
```
Comment (dramich) [inline pkg/controllers/management/k3supgrade/register.go]: This isn't used in this file, can it be moved?

File: pkg/controllers/management/k3supgrade/k3supgrade_handler.go [hunk 18]
```diff
@@ -0,0 +1,174 @@
...
+	if err != nil {
+		return err
+	}
+
+	appLister := userCtx.Management.Project.Apps("").Controller().Lister()
+	appClient := userCtx.Management.Project.Apps("")
+
+	latestVersionID := latestTemplateVersion.ExternalID
+
+	app, err := appLister.Get(systemProjectName, "rancher-k3s-upgrader")
```
Comment (dramich) [inline pkg/controllers/management/k3supgrade/k3supgrade_handler.go]: "rancher-k3s-upgrader" can move into a const

File: pkg/controllers/management/k3supgrade/k3supgrade_handler.go [hunk 19]
```diff
@@ -0,0 +1,174 @@
...
+	_, systemProjectName := ref.Parse(systemProjectID)
+
+	nsClient := userCtx.Core.Namespaces("")
+	appProjectName, err := utils2.EnsureAppProjectName(nsClient, systemProjectName, clusterName, systemUpgradeNS, creator.Name)
+	if err != nil {
+		return err
+	}
+
+	appLister := userCtx.Management.Project.Apps("").Controller().Lister()
+	appClient := userCtx.Management.Project.Apps("")
```
Comment (dramich) [inline pkg/controllers/management/k3supgrade/k3supgrade_handler.go]: Why don't these live on handler?

File: pkg/controllers/management/k3supgrade/k3supgrade_handler.go [hunk 20]
```diff
@@ -0,0 +1,174 @@
...
+			Spec: v32.AppSpec{
+				Description:     "Upgrade controller for k3s clusters",
+				ExternalID:      latestVersionID,
+				ProjectName:     appProjectName,
+				TargetNamespace: systemUpgradeNS,
+			},
+		}
+
+		// k3s upgrader doesn't exist yet, so it will need to be created
+		if _, err = appClient.Create(desiredApp); err != nil {
```
Comment (dramich) [inline pkg/controllers/management/k3supgrade/k3supgrade_handler.go]: Check the error isn't an alreadyExists error to save from races

File: pkg/controllers/management/k3supgrade/k3supgrade_handler.go [hunk 21]
```diff
@@ -0,0 +1,174 @@
...
+			return err
+		}
+	} else {
```
Comment (dramich) [inline pkg/controllers/management/k3supgrade/k3supgrade_handler.go]: Else block isn't needed

Comment (dramich) [reply pkg/controllers/management/k3supgrade/k3supgrade_handler.go]: Eh guess it is if you don't return after the create...

File: pkg/controllers/management/k3supgrade/register.go [hunk 22]
```diff
@@ -0,0 +1,190 @@
...
+	}
+
+	// create a client for GETing Plans in the downstream cluster
+	planConfig, err := planClientset.NewForConfig(&clusterCtx.RESTConfig)
+	if err != nil {
+		return err
+	}
+	planClient := planConfig.Plans(metav1.NamespaceAll)
+
+	planList, err := planClient.List(metav1.ListOptions{})
```
Comment (dramich) [inline pkg/controllers/management/k3supgrade/register.go]: Ideally this client can be moved onto the user context so we don't need to create it here. In the future other things will probably need it.

File: pkg/controllers/management/k3supgrade/k3supgrade_handler.go (hunk 19 above)
Comment (rmweir) [reply pkg/controllers/management/k3supgrade/k3supgrade_handler.go]: because the handler on the management cluster but the app is being deployed in the downstream cluster, which will be cluster specific.

Comment (dramich) [reply pkg/controllers/management/k3supgrade/k3supgrade_handler.go]: That doesn't make sense. App is a rancher CRD and this pathing is calling management, which gets back to the management context. This is not doing anything with the downstream cluster.

File: pkg/controllers/management/k3supgrade/k3supgrade_handler.go (hunk 20 above)
Comment (rmweir) [reply pkg/controllers/management/k3supgrade/k3supgrade_handler.go]: if it receives an already exists error then the handler should work on the next attempt

File: pkg/api/customization/cluster/validator.go [hunk 23]
```diff
@@ -118,6 +125,58 @@ func (v *Validator) validateEnforcement(request *types.APIContext, data map[stri
...
+	if request.Method == http.MethodPost {
+		return nil
+	}
+
+	if spec.K3sConfig == nil {
+		// only applies to k3s clusters
+		return nil
+	}
+
+	// must wait for original spec version to be set
```
Comment (cjellick) [inline pkg/api/customization/cluster/validator.go]: this comment is weird. you arent waiting on anything; you're just ensuring the user didnt send a ""

Comment (cjellick) [reply pkg/api/customization/cluster/validator.go]: and more important, returning `upgradeNotReadyErr` is wrong in this case

File: pkg/controllers/management/k3supgrade/register.go [hunk 24]
```diff
@@ -0,0 +1,190 @@
...
+		systemAccountManager:   systemaccount.NewManager(mgmtCtx),
+		manager:                manager,
+	}
+
+	wContext.Mgmt.Cluster().OnChange(ctx, "k3s-upgrade-controller", h.onClusterChange)
+}
+
+// deployPlans creates a master and worker plan in the downstream cluster to instrument
+// the system-upgrade-controller in the downstream cluster
+func (h *handler) deployPlans(cluster *v3.Cluster) error {
```
Comment (cjellick) [inline pkg/controllers/management/k3supgrade/register.go]: weird that this function is in register.go

File: pkg/controllers/management/k3supgrade/template.go [hunk 25]
```diff
@@ -0,0 +1,121 @@
...
+		},
+		Upgrade: &planv1.ContainerSpec{
+			Image: upgradeImage,
+		},
+	},
+	Status: planv1.PlanStatus{},
+}
+
+func generateMasterPlan(version string, concurrency int) planv1.Plan {
+	masterPlan := genericPlan
```
Comment (cjellick) [inline pkg/controllers/management/k3supgrade/template.go]: does this copy the plan?

File: pkg/controllers/management/k3supgrade/template.go [hunk 26]
```diff
@@ -0,0 +1,121 @@
...
+func generateWorkerPlan(version string, concurrency int) planv1.Plan {
+	workerPlan := genericPlan
+	workerPlan.Name = k3sWorkerPlanName
+	workerPlan.Spec.Version = version
+	workerPlan.Spec.Concurrency = int64(concurrency)
+
+	// worker plans wait for master plans to complete
+	workerPlan.Spec.Prepare = &planv1.ContainerSpec{
+		// TODO: the image version doesn't matter here, needs something recent
+		Image: upgradeImage + ":" + "v1.17.3-k3s1",
```
Comment (cjellick) [inline pkg/controllers/management/k3supgrade/template.go]: is this going to go to production like this or do we have a beter solution?

**Comment by cjellick** [comment]
failures are unrelated. im merging

File: pkg/controllers/management/k3supgrade/template.go (hunk 26 above)
Comment (daxmc99) [reply pkg/controllers/management/k3supgrade/template.go]: Ideally this would be fixed in system-upgrade-controller here https://github.com/rancher/system-upgrade-controller/issues/41

File: pkg/controllers/management/k3supgrade/register.go (hunk 24 above)
Comment (daxmc99) [reply pkg/controllers/management/k3supgrade/register.go]: Should be moved
